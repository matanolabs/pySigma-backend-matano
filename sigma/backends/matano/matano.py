from typing import Union, Any, ClassVar, Dict, Optional, Tuple, Pattern, List

import os
import re
import sys
import yaml
import json
import black
import textwrap

from sigma.conversion.state import ConversionState
from sigma.conversion.deferred import DeferredQueryExpression
from sigma.rule import SigmaRule, SigmaDetection, SigmaDetectionItem, SigmaLogSource
from sigma.conversion.base import TextQueryBackend, ProcessingPipeline
from sigma.conditions import ConditionItem, ConditionAND, ConditionOR, ConditionNOT, ConditionFieldEqualsValueExpression
from sigma.types import SigmaCompareExpression
from sigma.exceptions import SigmaFeatureNotSupportedByBackendError

def snake_case(s: str) -> str:
    return "_".join(
        re.sub(
            "([A-Z][a-z]+)",
            r" \1",
            re.sub("([A-Z]+)", r" \1", s.replace("-", " ").replace(".", " ")),
        ).split()
    ).lower()

def mkdir_if_not_exists(path: str):
    if not os.path.exists(path):
        os.makedirs(path, exist_ok=True)

class MatanoPythonBackend(TextQueryBackend):
    """Matano Python Backend for Sigma"""

    # Operator precedence: tuple of Condition{AND,OR,NOT} in order of precedence.
    # The backend generates grouping if required
    precedence : ClassVar[Tuple[ConditionItem, ConditionItem, ConditionItem]] = (ConditionNOT, ConditionAND, ConditionOR)
    group_expression : ClassVar[str] = "({expr})"   # Expression for precedence override grouping as format string with {expr} placeholder

    # Generated query tokens
    token_separator : str = " "     # separator inserted between all boolean operators
    or_token : ClassVar[str] = "or"
    and_token : ClassVar[str] = "and"
    not_token : ClassVar[str] = "not"
    eq_token : ClassVar[str] = " == "  # Token inserted between field and value (without separator)

    # String output
    ## Fields
    ### Quoting
    field_quote : ClassVar[str] = "'"                               # Character used to quote field characters if field_quote_pattern matches (or not, depending on field_quote_pattern_negation). No field name quoting is done if not set.
    field_quote_pattern : ClassVar[Pattern] = re.compile("^\\w+$")   # Quote field names if this pattern (doesn't) matches, depending on field_quote_pattern_negation. Field name is always quoted if pattern is not set.
    field_quote_pattern_negation : ClassVar[bool] = True            # Negate field_quote_pattern result. Field name is quoted if pattern doesn't matches if set to True (default).

    ### Escaping
    field_escape : ClassVar[str] = "\\"               # Character to escape particular parts defined in field_escape_pattern.
    field_escape_quote : ClassVar[bool] = True        # Escape quote string defined in field_quote
    field_escape_pattern : ClassVar[Pattern] = re.compile("\\s")   # All matches of this pattern are prepended with the string contained in field_escape.

    ## Values
    str_quote       : ClassVar[str] = '"'     # string quoting character (added as escaping character)
    escape_char     : ClassVar[str] = "\\"    # Escaping character for special characrers inside string
    wildcard_multi  : ClassVar[str] = "*"     # Character used as multi-character wildcard
    wildcard_single : ClassVar[str] = "*"     # Character used as single-character wildcard
    add_escaped     : ClassVar[str] = "\\"    # Characters quoted in addition to wildcards and string quote
    filter_chars    : ClassVar[str] = ""      # Characters filtered
    bool_values     : ClassVar[Dict[bool, str]] = {   # Values to which boolean values are mapped.
        True: "True",
        False: "False",
    }

    # String matching operators. if none is appropriate eq_token is used.
    startswith_expression : ClassVar[str] = "({field} and {field}.startswith({value}))"
    endswith_expression   : ClassVar[str] = "({field} and {field}.endswith({value}))"
    contains_expression   : ClassVar[str] = "{value} in {field}"
    wildcard_match_expression : ClassVar[str] = "fnmatch({field}, {value})"      # Special expression if wildcards can't be matched with the eq_token operator

    # Regular expressions
    re_expression : ClassVar[str] = "re.match({regex}, {field})"  # Regular expression query as format string with placeholders {field} and {regex}
    re_escape_char : ClassVar[str] = "\\"               # Character used for escaping in regular expressions
    re_escape : ClassVar[Tuple[str]] = ()               # List of strings that are escaped

    # cidr expressions
    cidr_expression : ClassVar[str] = "cidrmatch({field}, '{value}')"    # CIDR expression query as format string with placeholders {field} = {value}

    # Numeric comparison operators
    compare_op_expression : ClassVar[str] = "{field}{operator}{value}"  # Compare operation query as format string with placeholders {field}, {operator} and {value}
    # Mapping between CompareOperators elements and strings used as replacement for {operator} in compare_op_expression
    compare_operators : ClassVar[Dict[SigmaCompareExpression.CompareOperators, str]] = {
        SigmaCompareExpression.CompareOperators.LT  : "<",
        SigmaCompareExpression.CompareOperators.LTE : "<=",
        SigmaCompareExpression.CompareOperators.GT  : ">",
        SigmaCompareExpression.CompareOperators.GTE : ">=",
    }

    # Null/None expressions
    field_null_expression : ClassVar[str] = "{field} is None"          # Expression for field has null value as format string with {field} placeholder for field name

    # Field value in list, e.g. "field in (value list)" or "field containsall (value list)"
    convert_or_as_in : ClassVar[bool] = True                     # Convert OR as in-expression
    convert_and_as_in : ClassVar[bool] = False                    # Convert AND as in-expression
    in_expressions_allow_wildcards : ClassVar[bool] = False       # Values in list can contain wildcards. If set to False (default) only plain values are converted into in-expressions.
    field_in_list_expression : ClassVar[str] = "{field} {op} ({list})"  # Expression for field in list of values as format string with placeholders {field}, {op} and {list}
    or_in_operator : ClassVar[str] = "in"               # Operator used to convert OR into in-expressions. Must be set if convert_or_as_in is set
    and_in_operator : ClassVar[str] = "contains-all"    # Operator used to convert AND into in-expressions. Must be set if convert_and_as_in is set
    list_separator : ClassVar[str] = ", "               # List element separator

    # Value not bound to a field
    unbound_value_str_expression : ClassVar[str] = '"{value}"'   # Expression for string value not bound to a field as format string with placeholder {value}
    unbound_value_num_expression : ClassVar[str] = '{value}'   # Expression for number value not bound to a field as format string with placeholder {value}
    # unbound_value_re_expression : ClassVar[str] = '_=~{value}'    # Expression for regular expression not bound to a field as format string with placeholder {value}

    # Query finalization: appending and concatenating deferred query part
    deferred_start : ClassVar[str] = "\n| "               # String used as separator between main query and deferred parts
    deferred_separator : ClassVar[str] = "\n| "           # String used to join multiple deferred query parts
    deferred_only_query : ClassVar[str] = "*"            # String used as query if final query only contains deferred expression
 
    field_quote = None

    def __init__(self, processing_pipeline: Optional[ProcessingPipeline] = None, collect_errors: bool = False, **kwargs):
        super().__init__(processing_pipeline, collect_errors, **kwargs)
        self.used_cidr = False

    def escape_and_quote_field(self, field_name : str) -> str:
        val = super().escape_and_quote_field(field_name)
        parts = val.split(".")
        ret = "record"
        for i in range(len(parts)):
            part = parts[i]
            if i == len(parts) - 1:
                ret += f".get('{part}')"
            else:
                ret += f".get('{part}', {{}})"
        return ret

    def convert_condition_field_eq_val_cidr(self, cond: ConditionFieldEqualsValueExpression, state: ConversionState) -> Union[str, DeferredQueryExpression]:
        self.used_cidr = True
        return self.cidr_expression.format(field=self.escape_and_quote_field(cond.field), value=str(cond.value.network))

    def convert_value_re(self, r, state: ConversionState):
        val = super().convert_value_re(r, state)
        return "r" + "'" + val + "'"
    
    def is_keywords_detection(self, item: Union[SigmaDetectionItem, SigmaDetection]):
        if isinstance(item, SigmaDetectionItem):
            return item.is_keyword()
        else:
            return any(self.is_keywords_detection(x) for x in item.detection_items)

    def convert_rule(self, rule: SigmaRule, output_format: Optional[str] = None) -> List[Any]:
        for detection in rule.detection.detections.values():
            if self.is_keywords_detection(detection):
                raise SigmaFeatureNotSupportedByBackendError("Backend does not support keywords.")

        return super().convert_rule(rule, output_format)

    def _format_query(self, query: str):
        final_query = """\
import re, ipaddress
from fnmatch import fnmatch
"""

        if self.used_cidr:
            final_query += """\
def cidrmatch(ip, cidr):
    return ipaddress.ip_address(ip) in ipaddress.ip_network(cidr)
"""

        final_query += f"""\
def detect(record):
{textwrap.indent("return (" + query + ")", "    ")}
"""
        return black.format_str(final_query, mode=black.FileMode(line_length=100))

    def _format_logsource(self, ls: SigmaLogSource) -> Optional[str]:
        ret = "_".join(
            s for s in (ls.product, ls.category, ls.service) if s is not None
        )
        return ret

    def finalize_query_default(self, rule: SigmaRule, query: str, index: int, state: ConversionState) -> str:
        return self._format_query(query)

    def finalize_output_default(self, queries: List[str]) -> str:
        return list(queries)

    def finalize_query_detection(self, rule: SigmaRule, query: str, index: int, state: ConversionState) -> Any:
        final_query = self._format_query(query)
        title = snake_case(rule.title)

        comment_values = {
            "description": rule.description,
            "id": str(rule.id),
            "level": str(rule.level),
            "status": str(rule.status),
            "author": rule.author,
            "date": str(rule.date),
            "references": rule.references,
        }
        comment_values = { k: v for k,v in comment_values.items() if v is not None }

        comment = yaml.dump(comment_values, indent=4)
        comment = textwrap.indent(comment, "# ")

        ret = {
            "title": title,
            "detection_content": final_query,
            "comment": comment,
            "log_source": self._format_logsource(rule.logsource),
        }
        return ret

    def finalize_output_detection(self, queries: List[Any]):
        "Outputs rules as detection directories for Matano directory"

        ret = []
        for query in queries:
            title = query["title"].replace("/", "_")
            detection_dir = os.path.join(os.getcwd(), title)
            mkdir_if_not_exists(detection_dir)
            with open(os.path.join(detection_dir, "detect.py"), "w") as det_f:
                det_f.write(query["detection_content"])

            log_sources = [query["log_source"]]

            detection_yml_content = f"""\
# This file was generated from a Sigma rule

{query["comment"]}

name: {json.dumps(query["title"])}
log_sources: {json.dumps(log_sources)}
"""

            with open(os.path.join(detection_dir, "detection.yml"), "w") as config_f:
                config_f.write(detection_yml_content)

            ret.append(detection_dir)

        return ret
