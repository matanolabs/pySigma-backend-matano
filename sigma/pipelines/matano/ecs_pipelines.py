from sigma.processing.pipeline import ProcessingPipeline, ProcessingItem
from sigma.processing.transformations import (
    FieldMappingTransformation,
    FieldPrefixMappingTransformation,
)

ecs_cloudtrail = ProcessingPipeline(
    vars={
        "json_load_prefixes": [
            "aws.cloudtrail.additional_eventdata",
            "aws.cloudtrail.request_parameters",
            "aws.cloudtrail.response_elements",
            "aws.cloudtrail.service_event_details",
        ]
    },
    items=[
        ProcessingItem(
            identifier="field_mapping",
            transformation=FieldMappingTransformation(
                {
                    "apiVersion": "aws.cloudtrail.api_version",
                    "awsRegion": "cloud.region",
                    "errorCode": "aws.cloudtrail.error_code",
                    "errorMessage": "aws.cloudtrail.error_message",
                    "eventID": "event.id",
                    "eventName": "event.action",
                    "eventSource": "event.provider",
                    "eventTime": "ts",
                    "eventType": "aws.cloudtrail.event_type",
                    "eventVersion": "aws.cloudtrail.event_version",
                    "managementEvent": "aws.cloudtrail.management_event",
                    "readOnly": "aws.cloudtrail.read_only",
                    "requestID": "aws.cloudtrail.request_id",
                    "resources.accountId": "aws.cloudtrail.resources.account_id",
                    "resources.ARN": "aws.cloudtrail.resources.arn",
                    "resources.type": "aws.cloudtrail.resources.type",
                    "sharedEventId": "aws.cloudtrail.shared_event_id",
                    "sourceIPAddress": "source.address",
                    "userAgent": "user_agent",
                    "userIdentity.accessKeyId": "aws.cloudtrail.user_identity.access_key_id",
                    "userIdentity.accountId": "cloud.account.id",
                    "userIdentity.arn": "aws.cloudtrail.user_identity.arn",
                    "userIdentity.invokedBy": "aws.cloudtrail.user_identity.invoked_by",
                    "userIdentity.principalId": "user.id",
                    "userIdentity.sessionContext.attributes.creationDate": "aws.cloudtrail.user_identity.session_context.creation_date",
                    "userIdentity.sessionContext.attributes.mfaAuthenticated": "aws.cloudtrail.user_identity.session_context.mfa_authenticated",
                    "userIdentity.sessionContext.sessionIssuer.userName": "role.name",
                    "userIdentity.type": "aws.cloudtrail.user_identity.type",
                    "userIdentity.userName": "user.name",
                    "vpcEndpointId": "aws.cloudtrail.vpc_endpoint_id",
                }
            ),
        ),
        ProcessingItem(
            identifier="field_mapping_prefix",
            transformation=FieldPrefixMappingTransformation(
                {
                    "additionalEventdata": "aws.cloudtrail.additional_eventdata",
                    "requestParameters": "aws.cloudtrail.request_parameters",
                    "responseElements": "aws.cloudtrail.response_elements",
                    "serviceEventDetails": "aws.cloudtrail.service_event_details",
                }
            ),
        ),
    ]
)
