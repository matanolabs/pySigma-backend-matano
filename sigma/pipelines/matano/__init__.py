from .ecs_pipelines import ecs_cloudtrail
