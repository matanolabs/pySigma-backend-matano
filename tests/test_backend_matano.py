import pytest
from sigma.exceptions import SigmaFeatureNotSupportedByBackendError
from sigma.collection import SigmaCollection
from sigma.backends.matano import MatanoPythonBackend

@pytest.fixture
def matano_backend():
    return MatanoPythonBackend()

def test_matano_and_expression(matano_backend : MatanoPythonBackend, snapshot):
    ret = matano_backend.convert(
        SigmaCollection.from_yaml("""
            title: Test
            status: test
            logsource:
                category: test_category
                product: test_product
            detection:
                sel:
                    fieldA: valueA
                    fieldB: valueB
                condition: sel
        """)
    )

    assert ret[0] == snapshot

def test_matano_or_expression(matano_backend : MatanoPythonBackend, snapshot):
    assert matano_backend.convert(
        SigmaCollection.from_yaml("""
            title: Test
            status: test
            logsource:
                category: test_category
                product: test_product
            detection:
                sel1:
                    fieldA: valueA
                sel2:
                    fieldB: valueB
                condition: 1 of sel*
        """)
    )[0] == snapshot

def test_matano_and_or_expression(matano_backend : MatanoPythonBackend, snapshot):
    assert matano_backend.convert(
        SigmaCollection.from_yaml("""
            title: Test
            status: test
            logsource:
                category: test_category
                product: test_product
            detection:
                sel:
                    fieldA:
                        - valueA1
                        - valueA2
                    fieldB:
                        - valueB1
                        - valueB2
                condition: sel
        """)
    )[0] == snapshot

def test_matano_or_and_expression(matano_backend : MatanoPythonBackend, snapshot):
    assert matano_backend.convert(
        SigmaCollection.from_yaml("""
            title: Test
            status: test
            logsource:
                category: test_category
                product: test_product
            detection:
                sel1:
                    fieldA: valueA1
                    fieldB: valueB1
                sel2:
                    fieldA: valueA2
                    fieldB: valueB2
                condition: 1 of sel*
        """)
    )[0] == snapshot

def test_matano_in_expression(matano_backend : MatanoPythonBackend, snapshot):
    assert matano_backend.convert(
        SigmaCollection.from_yaml("""
            title: Test
            status: test
            logsource:
                category: test_category
                product: test_product
            detection:
                sel:
                    fieldA:
                        - valueA
                        - valueB
                        - valueC*
                condition: sel
        """)
    )[0] == snapshot

def test_matano_regex_query(matano_backend : MatanoPythonBackend, snapshot):
    assert matano_backend.convert(
        SigmaCollection.from_yaml("""
            title: Test
            status: test
            logsource:
                category: test_category
                product: test_product
            detection:
                sel:
                    fieldA|re: foo.*bar
                    fieldB: foo
                condition: sel
        """)
    )[0] == snapshot

def test_matano_cidr_query(matano_backend : MatanoPythonBackend, snapshot):
    assert matano_backend.convert(
        SigmaCollection.from_yaml("""
            title: Test
            status: test
            logsource:
                category: test_category
                product: test_product
            detection:
                sel:
                    field|cidr: 192.168.0.0/16
                condition: sel
        """)
    )[0] == snapshot

def test_matano_field_name_with_whitespace(matano_backend : MatanoPythonBackend, snapshot):
    assert matano_backend.convert(
        SigmaCollection.from_yaml("""
            title: Test
            status: test
            logsource:
                category: test_category
                product: test_product
            detection:
                sel:
                    field name: value
                    fieldA|cidr: 
                        - 192.168.0.0/14
                        - 192.168.0.0/14
                condition: sel
        """)
    )[0] == snapshot

def test_cidr(matano_backend: MatanoPythonBackend, snapshot):
    r1 = SigmaCollection.from_yaml("""
title: Test
status: test
logsource:
    category: test_category
    product: test_product
detection:
    sel:
        fieldB|cidr: 192.168.0.0/14
        fieldA|cidr: 
            - 192.168.0.0/14
            - 192.168.0.0/14
    condition: sel
""")
    ret = matano_backend.convert(r1)
    assert ret[0] == snapshot

def test_wildcards(matano_backend: MatanoPythonBackend, snapshot):
    r1 = SigmaCollection.from_yaml("""
title: Test
status: test
logsource:
    category: test_category
    product: test_product
detection:
    selection:
        CommandLine:
            - '*netsh* wlan show profile*'
            - '*netsh wlan show profile*'
    condition: selection
""")
    ret = matano_backend.convert(r1)
    assert ret[0] == snapshot


def test_keywords(matano_backend: MatanoPythonBackend):
    rule = SigmaCollection.from_yaml("""
title: Connection Proxy
id: 72f4ab3f-787d-495d-a55d-68c2ff46cf4c
status: test
description: Detects setting proxy
author: Ömer Günal
references:
  - https://attack.mitre.org/techniques/T1090/
date: 2020/06/17
modified: 2021/11/27
logsource:
  product: linux
detection:
  keywords:
    - 'http_proxy=*'
    - 'https_proxy=*'
  condition: keywords
falsepositives:
  - Legitimate administration activities
level: low
tags:
  - attack.defense_evasion
""")
    with pytest.raises(SigmaFeatureNotSupportedByBackendError):
        matano_backend.convert(rule)

    rule2 = SigmaCollection.from_yaml("""
title: Test
status: test
logsource:
    category: test_category
    product: test_product
detection:
    execve:
        type: 'EXECVE'
    truncate:
        - 'truncate'
        - '-s'
    dd:
        - 'dd'
        - 'if='
    filter:
        - 'of='
    condition: execve and (all of truncate or (all of dd and not filter))
""")
    with pytest.raises(SigmaFeatureNotSupportedByBackendError):
        matano_backend.convert(rule2)

def test_output_detection(matano_backend: MatanoPythonBackend):
    r1 = SigmaCollection.from_yaml("""
title: Test
status: test
logsource:
    category: test_category
    product: test_product
detection:
    selection:
        CommandLine:
            - '*netsh* wlan show profile*'
            - '*netsh wlan show profile*'
    condition: selection
""")
    ret = matano_backend.convert(r1, "detection")
    print(ret)
