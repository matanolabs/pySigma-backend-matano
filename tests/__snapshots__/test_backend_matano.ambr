# name: test_cidr
  '''
  import re, ipaddress
  from fnmatch import fnmatch
  
  
  def cidrmatch(ip, cidr):
      return ipaddress.ip_address(ip) in ipaddress.ip_network(cidr)
  
  
  def detect(record):
      return cidrmatch(record.get("fieldB"), "192.168.0.0/14") and (
          cidrmatch(record.get("fieldA"), "192.168.0.0/14")
          or cidrmatch(record.get("fieldA"), "192.168.0.0/14")
      )
  
  '''
# ---
# name: test_matano_and_expression
  '''
  import re, ipaddress
  from fnmatch import fnmatch
  
  
  def detect(record):
      return record.get("fieldA") == "valueA" and record.get("fieldB") == "valueB"
  
  '''
# ---
# name: test_matano_and_or_expression
  '''
  import re, ipaddress
  from fnmatch import fnmatch
  
  
  def detect(record):
      return (record.get("fieldA") in ("valueA1", "valueA2")) and (
          record.get("fieldB") in ("valueB1", "valueB2")
      )
  
  '''
# ---
# name: test_matano_cidr_query
  '''
  import re, ipaddress
  from fnmatch import fnmatch
  
  
  def cidrmatch(ip, cidr):
      return ipaddress.ip_address(ip) in ipaddress.ip_network(cidr)
  
  
  def detect(record):
      return cidrmatch(record.get("field"), "192.168.0.0/16")
  
  '''
# ---
# name: test_matano_field_name_with_whitespace
  '''
  import re, ipaddress
  from fnmatch import fnmatch
  
  
  def cidrmatch(ip, cidr):
      return ipaddress.ip_address(ip) in ipaddress.ip_network(cidr)
  
  
  def detect(record):
      return record.get("field\ name") == "value" and (
          cidrmatch(record.get("fieldA"), "192.168.0.0/14")
          or cidrmatch(record.get("fieldA"), "192.168.0.0/14")
      )
  
  '''
# ---
# name: test_matano_in_expression
  '''
  import re, ipaddress
  from fnmatch import fnmatch
  
  
  def detect(record):
      return (
          record.get("fieldA") == "valueA"
          or record.get("fieldA") == "valueB"
          or record.get("fieldA").startswith("valueC")
      )
  
  '''
# ---
# name: test_matano_or_and_expression
  '''
  import re, ipaddress
  from fnmatch import fnmatch
  
  
  def detect(record):
      return (
          record.get("fieldA") == "valueA1"
          and record.get("fieldB") == "valueB1"
          or record.get("fieldA") == "valueA2"
          and record.get("fieldB") == "valueB2"
      )
  
  '''
# ---
# name: test_matano_or_expression
  '''
  import re, ipaddress
  from fnmatch import fnmatch
  
  
  def detect(record):
      return record.get("fieldA") == "valueA" or record.get("fieldB") == "valueB"
  
  '''
# ---
# name: test_matano_regex_query
  '''
  import re, ipaddress
  from fnmatch import fnmatch
  
  
  def detect(record):
      return re.match(r"foo.*bar", record.get("fieldA")) and record.get("fieldB") == "foo"
  
  '''
# ---
# name: test_wildcards
  '''
  import re, ipaddress
  from fnmatch import fnmatch
  
  
  def detect(record):
      return fnmatch(record.get("CommandLine"), "*netsh* wlan show profile*") or record.get(
          "CommandLine"
      ).contains("netsh wlan show profile")
  
  '''
# ---
